### PR TITLE
[PyTorch] Refactor caching of cumulative sequence lengths

### PR DESCRIPTION
**Pull request by timmoon10**
To avoid redundantly calculating cumulative sequence lengths for attention, we only compute when `layer_number == 1` and otherwise use a cached value. However, pipeline parallelism breaks this optimization since most ranks will never have a layer with `layer_number == 1`. This PR removes the caching logic's dependency on `layer_number`, which unfortunately reintroduces redundant calculation except in cases where the sequence lengths are fixed.

This is a quick bugfix, but discussion is welcomed on how best to avoid the redundant calculation. Maybe we could split `layer_number` into two things: the local layer number and the scaling factor used at: https://github.com/NVIDIA/TransformerEngine/blob/6c1a8bb5dffbce386380f8e5a12c45f7032d9b76/transformer_engine/pytorch/attention.py#L1318

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -1671,58 +1691,45 @@ def forward(
...
-                            _cu_seqlens_kv = cu_seqlens_kv
-                            _indices_q = get_indices(max_seqlen_q, cu_seqlens_q)
-                            _indices_kv = get_indices(max_seqlen_kv, cu_seqlens_kv)
-                    query_layer_packed = PackTensors.apply(_indices_q, query_layer)
-                    key_layer_packed, value_layer_packed = PackTensors.apply(
-                        _indices_kv, key_layer, value_layer
+                    if cu_seqlens_q is None or cu_seqlens_kv is None:
+                        assert (attention_mask is not None
+                            ), "Please provide attention_mask for padding!"
+                        cu_seqlens_q, indices_q = get_cu_seqlens_and_indices(
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:1708]: Ok, so here we are basically not doing caching anymore, right? This makes sense I guess since it depends on the contents of the mask.

Comment (timmoon10) [reply transformer_engine/pytorch/attention.py:1708]: Yea, I think we'll need a new API to safely support this optimization.

**Comment by minitu** [comment]
Adding that MLPerf LLM training is currently using the `release_v1.3` branch because this change was merged in that release branch but not into main, so please push this forward.

**Review by ksivaman — APPROVED** [review]
LGTM.

**Comment by ksivaman** [comment]
I just saw @ptrendx's [comment](https://github.com/NVIDIA/TransformerEngine/pull/635#issue-2102615830) on #635 regarding the issue of pipeline parallelism and how `layer_number==1` is not set for any layer. To re-enable caching and remove this issue, wouldn't a simple fix be to assume/set layer number to 1 when not provided, or to always recompute the `cu_seqlens*` tensors and indices when layer number is not set?

@timmoon10

**Comment by ptrendx** [comment]
@ksivaman The problem is that NeMo is setting the layer number, but it is counting them "globally" in the full model which is cut using PP. This behavior makes sense so we should not assume that layer_number==1 will ever be true on a given GPU.

**Comment by parthmannan** [comment]
To add another observation here why layer_number based caching does not work - There are model implementations where attention is called multiple times in a transformer block with a different sequence length and batch size shape. Relying on layer number caching breaks this as the cache is set by the last attention to be called in layer_number=1. The next layer where the 1st attention is called, it ends up with an incorrect cu_seqlens_q/kv.
